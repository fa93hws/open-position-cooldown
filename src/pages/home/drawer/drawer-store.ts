import type {
  IntentionSchema,
  IntentionService,
} from '@services/intention/intention';
import { makeObservable, observable, action } from 'mobx';

export class DrawerStore {
  open = false;

  constructor(
    private intentionService: IntentionService,
    private afterSubmit: () => void,
  ) {
    makeObservable(this, {
      open: observable.ref,
      setOpen: action,
    });
  }

  setOpen(value: boolean) {
    this.open = value;
  }

  submitIntention = async (intention: IntentionSchema) => {
    await this.intentionService.addIntention(intention);
    this.setOpen(false);
    this.afterSubmit();
  };
}
