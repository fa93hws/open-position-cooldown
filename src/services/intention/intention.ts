import { LocalStorageService } from '@services/local-storage/local-storage';

export type IntentionSchema = {
  basicInfo: {
    name: string;
    code: string;
    market: string;
    // yyyy-MM-dd
    date: string;
  };
  // reason for holding the position
  reasons: string[];
  // why the current price is good
  priceExplain: string;
  strategy: {
    initial: {
      price: number;
      quantity: number;
    };
    plan: {
      price: number;
      quantity: number;
    }[];
    // say shit when droping to this price
    shitPrice: number;
  };
};

export class IntentionService {
  constructor(private readonly localStorageService: LocalStorageService) {}

  async addIntention(intention: IntentionSchema) {
    this.localStorageService.addToArray('intentions', intention);
  }
}
