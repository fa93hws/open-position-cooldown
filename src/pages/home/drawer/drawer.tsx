import * as React from 'react';
import { Drawer, Box, AppBar, Typography, IconButton } from '@material-ui/core';
import { makeStyles, withTheme, WithTheme } from '@material-ui/core/styles';
import { HighlightOff } from '@material-ui/icons';
import { observer } from 'mobx-react';

import type { IntentionService } from '@services/intention/intention';
import { sizes } from '@styles/styles';
import { createForm } from './form/form';
import { DrawerStore } from './drawer-store';

const useStyles = makeStyles((theme) => ({
  headContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-around',
    height: sizes[3],
    [theme.breakpoints.up('sm')]: {
      height: sizes[4],
    },
  },
  iconWrapper: {
    color: theme.palette.primary.contrastText,
  },
  placeholderIconWrapper: {
    color: 'transparent',
    userSelect: 'none',
    cursor: 'default',
  },
}));

type HomeDrawerProps = WithTheme & {
  open: boolean;
  onClose(): void;
  Form: React.ComponentType;
};

export const HomeDrawer = React.memo(
  withTheme((props: HomeDrawerProps) => {
    const styles = useStyles();
    return (
      <Drawer anchor="bottom" open={props.open}>
        <Box height="100vh" width="100%" display="flex" flexDirection="column">
          <AppBar className={styles.headContainer} position="static">
            {/* placeholder to balance the layout */}
            <IconButton className={styles.placeholderIconWrapper}>
              <HighlightOff fontSize="large" />
            </IconButton>
            <Typography variant="h4" component="h1">
              真的要买？
            </Typography>
            <IconButton className={styles.iconWrapper} onClick={props.onClose}>
              <HighlightOff fontSize="large" />
            </IconButton>
          </AppBar>
          <Box flex={1}>
            <props.Form />
          </Box>
        </Box>
      </Drawer>
    );
  }),
);

export function createHomeDrawer(
  intentionService: IntentionService,
  afterSubmit: () => void,
) {
  const store = new DrawerStore(intentionService, afterSubmit);
  const handleClose = () => store.setOpen(false);
  const showDrawer = () => store.setOpen(true);

  const Form = createForm(store.submitIntention);
  const Component = observer(() => (
    <HomeDrawer open={store.open} onClose={handleClose} Form={Form} />
  ));
  return {
    Component,
    showDrawer,
  };
}
