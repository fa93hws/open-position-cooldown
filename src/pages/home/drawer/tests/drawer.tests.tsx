import * as React from 'react';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';

import { IntentionService } from '@services/intention/intention';
import { createHomeDrawer, HomeDrawer } from '../drawer';

describe('HomeDrawer', () => {
  const Form = () => <div>form</div>;
  it('renders an open drawer', () => {
    expect(
      render(<HomeDrawer open onClose={jest.fn()} Form={Form} />).baseElement,
    ).toMatchSnapshot();
  });

  it('renders a closed drawer', () => {
    expect(
      render(<HomeDrawer open onClose={jest.fn()} Form={Form} />).baseElement,
    ).toMatchSnapshot();
  });

  it('can be mounted', () => {
    const intentionService = new IntentionService();
    const { Component } = createHomeDrawer(intentionService);
    expect(() => shallow(<Component />)).not.toThrow();
  });
});
