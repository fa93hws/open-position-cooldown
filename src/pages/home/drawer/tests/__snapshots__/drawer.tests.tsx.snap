// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeDrawer renders a closed drawer 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDrawer-root MuiDrawer-modal"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiDrawer-paper MuiDrawer-paperAnchorBottom MuiPaper-elevation16"
      tabindex="-1"
    >
      <div
        class="MuiBox-root MuiBox-root-9"
      >
        <header
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-headContainer-6 MuiPaper-elevation4"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-placeholderIconWrapper-8"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14.59 8L12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <h1
            class="MuiTypography-root MuiTypography-h4"
          >
            真的要买？
          </h1>
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-iconWrapper-7"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14.59 8L12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </header>
        <div
          class="MuiBox-root MuiBox-root-10"
        >
          <div>
            form
          </div>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`HomeDrawer renders an open drawer 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDrawer-root MuiDrawer-modal"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiDrawer-paper MuiDrawer-paperAnchorBottom MuiPaper-elevation16"
      tabindex="-1"
    >
      <div
        class="MuiBox-root MuiBox-root-4"
      >
        <header
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-headContainer-1 MuiPaper-elevation4"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-placeholderIconWrapper-3"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14.59 8L12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <h1
            class="MuiTypography-root MuiTypography-h4"
          >
            真的要买？
          </h1>
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-iconWrapper-2"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14.59 8L12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </header>
        <div
          class="MuiBox-root MuiBox-root-5"
        >
          <div>
            form
          </div>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
